### PR TITLE
feat: Initial oathkeeper charm

### PR DESCRIPTION
**Pull request by natalian98**
## Issue

Oathkeeper is currently not charmed and only available as a rock and docker image.

## Solution

This PR adds implementation of the initial charm:
- config and sample access rule files
- pebble layer
- events observation
- traefik-k8s ingress relation
- unit and integration tests.

## Additional context

We are not using the 4455 reverse proxy port, but it's always exposed by oathkeeper, even if you omit the `proxy` part in config (see the [dockerfile](https://github.com/ory/oathkeeper/blob/master/Dockerfile#L16)). You can verify it by ssh into the pod and running `netstat -tulpn`. That port will need to be protected with a k8s network policy to deny the traffic, I'll open another PR for that.

Note that some values, like kratos url or access rules, are placeholders/dummy values and will be replaced by juju integrations.

## Testing Instructions

Deploy the charm:
`juju deploy ./oathkeeper*.charm --resource oci-image=$(yq eval '.resources.oci-image.upstream-source' metadata.yaml)`
Find oathkeeper's k8s svc address and curl the ready endpoint:
`curl -v <svc-address>:4456/health/ready`

It should return `{"status":"ok"}`

`curl -v <svc-address>:4456/rules` should return a sample access rule.

## Release Notes

initial charm implementation

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -8,18 +8,157 @@
...
+    @property
+    def _oathkeeper_service_is_created(self) -> bool:
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    @property
+    def _kratos_login_url(self) -> str:
```
Comment (wood-push-melon) [inline src/charm.py:92]: I am still new to the charming, just a tiny question here:

- Why we don't initialize `_kratos_login_url` in the `__init__` method?

File: src/charm.py [hunk 2]
```diff
@@ -8,18 +8,157 @@
...
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring resources")
+
+        self._container.add_layer(self._container_name, self._oathkeeper_layer, combine=True)
+
+        if not self._oathkeeper_service_is_created:
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Oathkeeper service")
```
Comment (wood-push-melon) [inline src/charm.py:132]: Nitpick from a charming freshman:

- Like above, `Waiting for Oathkeeper service` to be `created` or `ready`?

File: src/charm.py [hunk 3]
```diff
@@ -8,18 +8,157 @@
...
+
+    def _on_oathkeeper_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Event Handler for pebble ready event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Oathkeeper container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring resources")
```
Comment (wood-push-melon) [inline src/charm.py:126]: Nitpick from a charming freshman:

-  Is it better if we provide a little bit more for the `message` argument? I am not very clear on what resources it is configuring.

File: src/charm.py [hunk 4]
```diff
@@ -8,18 +8,157 @@
...
+        )
+        return rendered
+
+    def _render_conf_file(self) -> str:
+        """Render the Oathkeeper configuration file."""
+        with open("templates/oathkeeper.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_session_url=f"http://kratos.{self.model.name}.svc.cluster.local:4433/sessions/whoami",
```
Comment (wood-push-melon) [inline src/charm.py:113]: Nitpick from a charming freshman:

- Is it better we group the two Kratos' exposed endpoints together somewhere in the `__init__` method?

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -1,15 +1,160 @@
...
+
+    assert harness.charm.unit.status == WaitingStatus("Waiting to connect to Oathkeeper container")
+
+
+def test_update_container_config(harness: Harness) -> None:
+    harness.set_can_connect(CONTAINER_NAME, True)
+
+    harness.charm.on.oathkeeper_pebble_ready.emit(CONTAINER_NAME)
+
+    expected_config = {
```
Comment (wood-push-melon) [inline tests/unit/test_charm.py:49]: Have a tiny question about the necessity to construct the expected Oathkeeper configuration in this way. In this test case, I think what we would like to make sure is that the charm actually reads the Jinja template file and renders the configuration file with the correct parameters.

The inputs to the method `_render_conf_file` actually would be the template file and the two Kratos' parameters. So is it better that we use the expected template file and expected Kratos' parameters to render the configuration file as the `expected_config` instead of pasting the whole rendered configuration?

File: templates/access-rules.yaml.j2 [hunk 6]
```diff
@@ -0,0 +1,16 @@
+- id: "sample-rule:protected"
```
Comment (wood-push-melon) [inline templates/access-rules.yaml.j2:1]: As a freshman in Oathkeeper, I would like to learn:

- For the Oathkeeper's access rules and configuration, do we allow the client user to provide their own access rules and configuration file?
- If this is the case, how do we handle this case? e.g. we try to do a merge of the built-in and user-provided?

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -1,3 +1,99 @@
...
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    address = await get_app_address(ops_test, TRAEFIK)
+    health_check_url = f"http://{address}/{ops_test.model.name}-{APP_NAME}/health/ready"
+    resp = requests.get(health_check_url)
```
Comment (wood-push-melon) [inline tests/integration/test_charm.py:97]: I guess the health check request might not guarantee the scaling up is a success or failure. The `wait_for_model` already tells us the status after the scaling up.

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -1,3 +1,99 @@
...
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
+            raise_on_blocked=False,
+            status="active",
+            timeout=1000,
+        )
+        assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+
+
+async def test_ingress_relation(ops_test: OpsTest) -> None:
```
Comment (wood-push-melon) [inline tests/integration/test_charm.py:56]: A tiny question from a charming freshman:

- For the integration test, just wondering if is it necessary to test some negative cases, e.g. how the charm behaves after a relation is broken, etc.

File: src/charm.py (hunk 1 above)
Comment (natalian98) [reply src/charm.py]: I'm fine with initializing it in `init` for now, in (near) future there will be a method to take those urls from a kratos-oathkeeper relation interface.

File: tests/unit/test_charm.py (hunk 5 above)
Comment (natalian98) [reply tests/unit/test_charm.py:33]: It's a common pattern in our tests to check for the full rendered configuration file. In (near) future there will be more parameters.
That said, I will modify this test to check the `expected_config` against what was actually pushed to the container (use `container_config.read()`) to not rely on the `_render_conf_file` method.

File: templates/access-rules.yaml.j2 (hunk 6 above)
Comment (natalian98) [reply templates/access-rules.yaml.j2:1]: We will not allow the users to provide their own configuration file (that would be a bit against the idea of operators - charms should encapsulate the necessary logic and configuration), but we will enable admins to provide the access rules.
There will be a relation interface to implement the logic behind it, you can see some context in the Oathkeeper spec. We want to have only one source of the access rules per each app. There will be no built-in rules by default - this one is a sample for testing purposes.

File: tests/integration/test_charm.py (hunk 8 above)
Comment (natalian98) [reply tests/integration/test_charm.py]: Since I removed the ingress relation as suggested, there are no more relations to be tested for now - but sure, it makes sense to test relations removal :+1:

File: templates/access-rules.yaml.j2 (hunk 6 above)
Comment (wood-push-melon) [reply templates/access-rules.yaml.j2:1]: ic~~ thank you for the clarifications.

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+that you also need to add `lightkube` and `lightkube-models` to your charm's `requirements.txt`.**
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.observability_libs.v0.kubernetes_service_patch
+echo <<-EOF >> requirements.txt
+lightkube
+lightkube-models
+EOF
+```
```
Comment (shipperizer) [inline lib/charms/observability_libs/v0/kubernetes_service_patch.py:40]: random question about this, but all this docstring is not really related to this file, it's just boilerplate....do we need it in every charm code, is it standard practice or is is simply a leftover?

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+        client = Client()
+        try:
+            if self.service_name != self._app:
+                self._delete_and_create_service(client)
+            client.patch(Service, self.service_name, self.service, patch_type=PatchType.MERGE)
+        except ApiError as e:
+            if e.status.code == 403:
+                logger.error("Kubernetes service patch failed: `juju trust` this application.")
+            else:
+                logger.error("Kubernetes service patch failed: %s", str(e))
```
Comment (shipperizer) [inline lib/charms/observability_libs/v0/kubernetes_service_patch.py:237]:
```suggestion
                logger.error("Kubernetes service patch failed: %s", e)
```

do we need the string casting?

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    def _patch(self, _) -> None:
+        """Patch the Kubernetes service created by Juju to map the correct port.
+
+        Raises:
+            PatchFailed: if patching fails due to lack of permissions, or otherwise.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        client = Client()
```
Comment (shipperizer) [inline lib/charms/observability_libs/v0/kubernetes_service_patch.py:228]: is it not common practice to have the client as instance attribute?that would spare a new connection

File: src/charm.py [hunk 12]
```diff
@@ -8,18 +8,136 @@
...
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Oathkeeper Operator layer",
+                    "command": f"oathkeeper serve -c {self._oathkeeper_config_path}",
```
Comment (shipperizer) [inline src/charm.py:58]: any preference between https://github.com/canonical/oathkeeper-operator/pull/7/files#diff-6dee06b6a3ec89348da36c2c246178117307d18db2972e78809ad2d0bed7ceb8R235 and the current `f"` interpolation?

can we stick with one or is it ok to use different ways?

File: src/charm.py [hunk 13]
```diff
@@ -8,18 +8,136 @@
...
+        """Event Handler for pebble ready event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Oathkeeper container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring the container")
+
+        self._container.add_layer(self._container_name, self._oathkeeper_layer, combine=True)
```
Comment (shipperizer) [inline src/charm.py:115]: dummy question, [here](https://github.com/canonical/oathkeeper-operator/pull/7/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR118) we call \event.defer()\ and then we return 
i expect the code to flow back to this function and re add the layer, what does that imply?aka can we keep adding the very same layer?or we do it because the `self._container` gets reset due to the charm  being rerun from scratch?

(forgive me, forward me to the docs without any problem, just need some extra context really)

File: src/charm.py [hunk 14]
```diff
@@ -8,18 +8,136 @@
...
+            self._oathkeeper_access_rules_path, self._render_access_rules_file(), make_dirs=True
+        )
+        self._container.push(
+            self._oathkeeper_config_path, self._render_conf_file(), make_dirs=True
+        )
+
+        try:
+            self._container.start(self._container_name)
+        except ChangeError as err:
+            logger.error(str(err))
```
Comment (shipperizer) [inline src/charm.py:133]:
```suggestion
            logger.error(err)
```

do we need the casting?

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py (hunk 9 above)
Comment (natalian98) [reply lib/charms/observability_libs/v0/kubernetes_service_patch.py:40]: This lib is imported using [fetch-lib](https://juju.is/docs/sdk/charmcraft-fetch-lib), it's owned by the observability team. The docstring is required for all charmcraft libs.

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py (hunk 11 above)
Comment (natalian98) [reply lib/charms/observability_libs/v0/kubernetes_service_patch.py:228]: This library is owned by Observability team and just imported to our charms. You can open an issue in [observability libs](https://github.com/canonical/observability-libs/tree/main) repo.

File: src/charm.py (hunk 14 above)
Comment (natalian98) [reply src/charm.py:133]: Yes, I can't find the related issue but the errors produced by juju need to be casted to string.

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py (hunk 9 above)
Comment (nsklikas) [reply lib/charms/observability_libs/v0/kubernetes_service_patch.py:40]: This is used to populate the charmhub docs (https://charmhub.io/observability-libs/libraries/kubernetes_service_patch)

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py (hunk 11 above)
Comment (nsklikas) [reply lib/charms/observability_libs/v0/kubernetes_service_patch.py:228]: I suppose that since the charm code is initialized every time a hook is fired, it makes no difference if this is an attribute or not (it will actually save you a couple of connections since it will be always initialized only if it's needed).

File: src/charm.py [hunk 15]
```diff
@@ -8,18 +8,136 @@
...
+        )
+        return rendered
+
+    def _on_oathkeeper_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Event Handler for pebble ready event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Oathkeeper container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
```
Comment (nsklikas) [inline src/charm.py:111]: Why do we need to check this? If the pebble ready event is fired is it not safe to assume that we can connect to the container?

File: src/charm.py [hunk 16]
```diff
@@ -8,18 +8,136 @@
...
+        self.framework.observe(self.on.oathkeeper_pebble_ready, self._on_oathkeeper_pebble_ready)
+
+    @property
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
```
Comment (nsklikas) [inline src/charm.py:55]: Let's not use the `_container_name` for the service name (even if it has the same value let's use a different variable to avoid confusion)

File: src/charm.py [hunk 17]
```diff
@@ -8,18 +8,136 @@
...
+
+        self.unit.status = MaintenanceStatus("Configuring the container")
+
+        self._container.add_layer(self._container_name, self._oathkeeper_layer, combine=True)
+
+        if not self._oathkeeper_service_is_created:
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Oathkeeper service to be created")
+            logger.info("Oathkeeper service is absent. Deferring the event.")
+            return
```
Comment (nsklikas) [inline src/charm.py:121]: Since you add the layer in the previous line, can this `if` clause ever succeed?

File: tests/unit/conftest.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    harness.begin()
+    yield harness
+    harness.cleanup()
+
+
+@pytest.fixture()
+def mocked_kubernetes_service_patcher(mocker: MockerFixture) -> Generator:
+    mocked_service_patcher = mocker.patch("charm.KubernetesServicePatch")
+    mocked_service_patcher.return_value = lambda x, y: None
+    yield mocked_service_patcher
```
Comment (nsklikas) [inline tests/unit/conftest.py:27]: You can `return` instead of `yield`

File: tests/integration/test_charm.py [hunk 19]
```diff
@@ -1,3 +1,68 @@
...
+    """Check that oathkeeper works after it is scaled up."""
+    app = ops_test.model.applications[APP_NAME]
+
+    await app.scale(2)
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        wait_for_units=2,
```
Comment (nsklikas) [inline tests/integration/test_charm.py:63]: I think this is not needed, AFAIK without the `wait_for_units` `wait_for_idle` will wait until all units are active.

File: tests/integration/test_charm.py [hunk 20]
```diff
@@ -1,3 +1,68 @@
...
+
+    await ops_test.model.deploy(
+        application_name=APP_NAME,
+        entity_url=charm,
+        resources={"oci-image": oathkeeper_image_path},
+        series="jammy",
+        trust=True,
+    )
+
+    async with ops_test.fast_forward():
```
Comment (nsklikas) [inline tests/integration/test_charm.py:44]: I think that this is not needed (our charms should not rely on the update_status event to become active)

**Review by nsklikas — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 15 above)
Comment (natalian98) [reply src/charm.py:111]: According to [these](https://juju.is/docs/sdk/container-name-pebble-ready-event) docs:

> it’s essential to, also in pebble-ready event handlers, to guard with Container.can_connect() to check that the socket is up

At the time the pebble event is fired it should be able to connect to the container, but since the event can be deferred, I think it makes sense to check it.

File: src/charm.py (hunk 12 above)
Comment (natalian98) [reply src/charm.py:58]: There is no company-wide guideline to stick to any, but I think we should use f-strings as per PEP498

File: lib/charms/observability_libs/v0/kubernetes_service_patch.py (hunk 11 above)
Comment (shipperizer) [reply lib/charms/observability_libs/v0/kubernetes_service_patch.py:228]: so we have to version this code even though we don't own it?
not sure i follow, but basically if it comes from someone else then fine not to touch it, but if instead we own it i'd say we would be better not instantiating new stuff if not needed and recycle the current connections to the k8s api 

anyway not a biggie

File: src/charm.py (hunk 17 above)
Comment (nsklikas) [reply src/charm.py:122]: I don't see that happening in the charms of other teams and I am not sure if it makes much sense, I would suggest to remove it and add it back in if we find experience any errors.

File: .github/workflows/tests.yaml [hunk 21]
```diff
@@ -34,6 +34,12 @@ jobs:
           juju-channel: 3.1
           bootstrap-options: '--agent-version=3.1.0'
 
+      # Workaround for https://github.com/canonical/charmcraft/issues/907
+      - name: Change permissions touch the charm file to modify mtime
+        run: |
+          chmod u+x src/charm.py
+          touch src/charm.py
+
```
Comment (nsklikas) [inline .github/workflows/tests.yaml:42]: If you add the necessary permissions to `src/charm.py` and push them wouldn't this be solved?

**Review by nsklikas — APPROVED** [review]
